proposals/improved-interpolated-strings.md

```diff
@@ -0,0 +1,436 @@
+1. `E` is a non-constant _interpolated\_string\_expression_, `C1` is an _implicit\_string\_builder\_conversion_, `T1` is an _applicable\_interpolated\_string\_builder\_type_, and `C2` is not an _implicit\_string\_builder\_conversion_, or
+2. `E` does not exactly match `T2` and at least one of the following holds:
+    * `E` exactly matches `T1` ([Exactly matching Expression](expressions.md#exactly-matching-expression))
+    * `T1` is a better conversion target than `T2` ([Better conversion target](expressions.md#better-conversion-target))
+
+This does mean that there are some potentially non-obvious overload resolution rules, depending on whether the interpolated string in question is a constant-expression or not. For example:
+
+```cs
+void Log(string s) { ... }
+void Log(TraceLoggerParamsBuilder p) { ... }
+
+Log($""); // Calls Log(string s), because $"" is a constant expression
+Log($"{"test"}"); // Calls Log(string s), because $"{"test"}" is a constant expression
+Log($"{1}"); // Calls Log(TraceLoggerParamsBuilder p), because $"{1}" is not a constant expression
+```
+
+This is introduced so that things that can simply be emitted as constants do so, and don't incur any overhead, while things that cannot be constant use the builder pattern.
+
+### InterpolatedStringBuilder and Usage
+
+We introduce a new type in `System.Runtime.CompilerServices`: `InterpolatedStringBuilder`. This is a ref struct with many of the same semantics as `ValueStringBuilder`,
+intended for direct use by the C# compiler. This struct would look approximately like this:
+
+```cs
+public ref struct InterpolatedStringBuilder
+{
+    public void GetInterpolatedStringBuilder(int baseLength, int formatHoleCount, out InterpolatedStringBuilder builder)
+        => builder = new InterpolatedStringBuilder(baseLength, formatHoleCount);
+
+    private char[] _array;
+    internal int _count;
+
+    private InterpolatedStringBuilder(int baseLength, int formatHoleCount)
+    {
+        _array = ArrayPool<char>.Shared.Rent(baseLength /* Or some calculation based on what we see on average for the length of format holes */);
+        _count = 0;
+    }
+    public string ToString()
+    {
+        string result = _array.AsSpan(0, _count).ToString();
+        ArrayPool<char>.Shared.Return(_array);
+        Return result;
+    }
+    public bool TryFormat(string s) => TryFormat((ReadOnlySpan<char>)s);
+    public bool TryFormat(ReadOnlySpan<char> s)
+    {
+        if (s.Length >= _array.Length - _count) Grow();
+        s.AsSpan().CopyTo(_array);
+        _count += s.Length;
+        return true;
+    }
+    … // other TryFormat overloads for other types (including ReadOnlySpan<T>), a generic, etc.
+}
+```
+
+We also provide a new `string.Format` overload, as follows:
+
+```cs
+public class String
+{
+    public static string Format(InterpolatedStringBuilder builder) => builder.ToString();
```

Comment:
I think we'll want to bury this instead as a static on InterpolatedStringBuilder.

Reply:
We should also consider allowing such builder arguments to be `ref`, so that the implementation of the method can clear out the call site's reference, if it, for example, contains a pooled array or some expensive resource that should be cleared to avoid lifetime extension.